File: javascript/src/rateLimiter.js
'use strict';

/**
 * Process-local, fixed-window rate limiter store. Does NOT share state
 * across processes/instances -- see THREAT_MODEL.md.
 */
class MemoryRateLimiterStore {
  constructor() {
    this.buckets = new Map();
  }

  increment(key, windowMs) {
    const now = Date.now();
    let bucket = this.buckets.get(key);
    if (!bucket || now > bucket.resetAt) {
      bucket = { count: 0, resetAt: now + windowMs };
      this.buckets.set(key, bucket);
    }
    bucket.count += 1;
    return bucket.count;
  }
}

/**
 * Enforces a maximum number of actions per key within a rolling window.
 *
 * allow() is always async (returns a Promise<boolean>), even though
 * MemoryRateLimiterStore's own increment() is synchronous -- awaiting a
 * non-Promise value resolves it immediately, so this costs nothing for the
 * in-memory case, but it's required for any real network-backed store
 * (e.g. RedisRateLimiterStore), which cannot be synchronous. An earlier
 * version of this class was fully synchronous, which made it impossible
 * to implement a correct Redis-backed store at all (comparing a Promise
 * to limit with <= silently produces the wrong answer every time).
 */
class RateLimiter {
  constructor(store, limit, windowMs) {
    this.store = store;
    this.limit = limit;
    this.windowMs = windowMs;
  }

  async allow(key) {
    const count = await this.store.increment(key, this.windowMs);
    return count <= this.limit;
  }
}

module.exports = { RateLimiter, MemoryRateLimiterStore };

File: python/securekit/rate_limiter.py
"""Fixed-window rate limiting with a pluggable storage backend."""

import threading
import time
from typing import Protocol


class RateLimiterStore(Protocol):
    def increment(self, key: str, window_seconds: float) -> int:
        """Increment the counter for key, creating the window if absent,
        and return the new count."""
        ...


class MemoryRateLimiterStore:
    """Process-local, thread-safe fixed-window store. Does NOT share state
    across processes/instances -- see THREAT_MODEL.md."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._buckets: dict[str, tuple[int, float]] = {}

    def increment(self, key: str, window_seconds: float) -> int:
        with self._lock:
            now = time.time()
            count, reset_at = self._buckets.get(key, (0, 0.0))
            if now > reset_at:
                count, reset_at = 0, now + window_seconds
            count += 1
            self._buckets[key] = (count, reset_at)
            return count


class RateLimiter:
    def __init__(self, store: RateLimiterStore, limit: int, window_seconds: float) -> None:
        self._store = store
        self._limit = limit
        self._window_seconds = window_seconds

    def allow(self, key: str) -> bool:
        return self._store.increment(key, self._window_seconds) <= self._limit
